### ISSUE TITLE
Making the Ocean's boundary condition interface more flexible

### ISSUE THREAD

**Issue by glwagner** [CliMA/ClimateMachine.jl#1740]
Currently one specifies `KinematicStress` on a boundary in the `HydrostaticBoussinesqSuperModel` by using the type `Impenetrable{KinematicStress}` as a boundary condition and writing a function `kinematic_stress` that dispatches on `problem`:

https://github.com/CliMA/ClimateMachine.jl/blob/25f4421449c0e3f10366b7ffb540992127301101/src/Ocean/HydrostaticBoussinesq/bc_velocity.jl#L233

This approach has a few limitations:

* We want users to be able to specify boundary conditions (Flux, Gradient, Value) by providing either numbers or functions. We don't want to require users to define new types.

* A user can only specify one `kinematic_stress` per problem type. Yet in principle one could impose a different kinematic stress condition on each boundary (of which there are 6, for example, in a `RectangularDomain`)

* We want to support boundary conditions that are arbitrary functions of the spatial coordinate, but only `y` is currently supported.

**Comment by glwagner** [comment]
cc @blallen @christophernhill @jm-c

**Comment by glwagner** [comment]
As a relatively quick solution we can solve this by 

1. Adding `x, y, z` as auxiliary variables (hopefully in the future we could figure out how to infer `x, y, z` without this) and

2. Redefining `KinematicStress` to

```julia
struct KinematicStress{B, S} <: VelocityDragBC
    stress :: S

    function KinematicStress{B}(stress::S) where {B, S}
        new{B, S}(stress)
    end
end
```

We then need

```julia
@inline function ocean_velocity_boundary_state!(
    nf::NumericalFluxSecondOrder,
    bc_velocity::Impenetrable{KinematicStress},
    ocean,
    Q⁺,
    D⁺,
    A⁺,
    n⁻,
    Q⁻,
    D⁻,
    A⁻,
    t,
)
    Q⁺.u = Q⁻.u
    D⁺.ν∇u = n⁻ * getbc(bc_velocity.drag, A⁻.x, A⁻.y, A⁻.z, t)

    return nothing
end

getbc(bc::KinematicStress{B, <:Tuple{<:Number, <:Number}}, x, y, z, t) where B = bc.stress

getbc(bc::KinematicStress{:x}, x, y, z, t) = bc.stress(y, z, t)
getbc(bc::KinematicStress{:y}, x, y, z, t) = bc.stress(x, z, t)
getbc(bc::KinematicStress{:z}, x, y, z, t) = bc.stress(x, y, t)
```

One then constructs a `KinematicStress` object with a function that returns the same object that `kinematic_stress(ocean.problem, A⁻.y, ocean.ρₒ)` would have returned.

As a side note, I think it is inconvenient that users need to specify their stress as a vector here. I think it would be more convenient to specify the components of the stress vector independently.

**Comment by glwagner** [comment]
@christophernhill suggests that an alternative method for diagnosing the boundary you're currently on would be to rewrite this function:

https://github.com/CliMA/ClimateMachine.jl/blob/84caa94eb1bfb7ffd48cef73a8a400b70fc7d87f/src/Ocean/OceanBC.jl#L97-L130

I'm not sure how to do that; I would need someone to explain to me what this function is doing.

**Comment by christophernhill** [comment]
> @christophernhill suggests that an alternative method for diagnosing the boundary you're currently on would be to rewrite this function:
> 
> https://github.com/CliMA/ClimateMachine.jl/blob/84caa94eb1bfb7ffd48cef73a8a400b70fc7d87f/src/Ocean/OceanBC.jl#L97-L130
> 
> I'm not sure how to do that; I would need someone to explain to me what this function is doing.

I think this already does what we need with KinematicStress type holding a function. 

Can't I then define, for example (and hypothetically), 

```
wind_stress = F1(x,y,z,t,p)
side_wall_stress = F2(x,y,z,p)
```

and

```
    boundary_conditions = (
      OceanBC(Impenetrable(KinematicStress(stress=side_wall_stress)), Insulating()),
      OceanBC(Impenetrable(FreeSlip()), Insulating()),
      OceanBC(Penetrable(KinematicStress(stress=wind_stress)), Insulating()),
    ),

```

I still have to be careful about normal vectors, directions etc... but in principle that can be done.
I can use bc id numbers as I see fit. 
For example I can have 4 different bc numbers for north, south, west, east etc... in a simple cartesian case and four functions or variations on a common
function with a flag.  The different functions would show up as the stress() member of the KStress type instance....

@simonbyrne understands the @generated loop - but I think that is what should be possible. 

I don't think the getbc foo should be needed?

**Comment by christophernhill** [comment]
we can also write a generic components_to_tensor function I think to fill out
```
D⁺.ν∇u
```
etc....  from generic stress components τ¹¹, τ²² relative to the local boundary. Might be best to start with simple case - rectangular, orthogonal grid and only diagonals for only tangential to wall vectors.

**Comment by glwagner** [comment]
You're right, we don't need `getbc`, provided users are okay with writing boundary conditions as functions of `x, y, z`.

As for parameters, we should use callable objects for that. My experience is that its better to carry around local references to variables, rather than trying to put them in some global store and pass them through a dense tree of functions. For now, using globally referenced variables (with `const` applied as appropriate on the GPU) is probably a fine short-term solution.

**Comment by glwagner** [comment]
> we can also write a generic components_to_tensor function I think to fill out

I don't know how to do that --- can you illustrate with code?

**Comment by christophernhill** [comment]
> > we can also write a generic components_to_tensor function I think to fill out
> 
> I don't know how to do that --- can you illustrate with code?

I think it would a bit like this? 

https://github.com/CliMA/ClimateMachine.jl/blob/84caa94eb1bfb7ffd48cef73a8a400b70fc7d87f/src/Ocean/OceanProblems/SimpleBoxProblem.jl#L380

**Comment by charleskawczynski** [comment]
One thing I’ve been pushing for is to develop a user interface without the +/- states— and have the +/- states determined in a separate layer. I also definitely agree with having a layer where we define Dirichlet, Neumann and Robin BCs.

**Comment by glwagner** [comment]
> I also definitely agree with having a layer where we define Dirichlet, Neumann and Robin BCs.

Robin is another kind of BC that I didn't mention. That'd be good too.

Specifying fluxes directly should also be an option, I think, though the precise application to DG methods is hazy to me. But as far as I can tell, specifying fluxes is the simplest boundary condition in the weak formulation of the equations (?) For large-scale ocean / atmospheric modeling (LES to GCM scales) it seems that this is in fact the only boundary condition we need. Idealized work does benefit from the ability to prescribe values (Dirichlet) and gradients (Neumann) on boundaries... so we should support those too.

**Comment by christophernhill** [comment]
So I think the form for stress could be along the lines of

```
   D⁺.ν∇u = n⁻ * (@SVector[ bc_velocity.stress.τ¹¹( x,y,z,t),bc_velocity.stress.τ²²(x,y,z,t)])'
```
with τ¹¹ and τ²² defined in the scripting layer and -0 by default

rather than

```
    D⁺.ν∇u = n⁻ * kinematic_stress(ocean.problem, A⁻.y, ocean.ρₒ)'
```

A couple of things I am unclear on - but don't stop coding

- its a little fuzzy to me why D⁺.ν∇u seems to be a 2 element vector and not a full gradient tensor thing. 
- we currently seem to set the bc in ```nf::NumericalFluxSecondOrder,```, which is because we are setting through a k.grad term at boundary. Thats OK, but we could equally have the k.grad term as zero I think and have some sort of source term? In the end we are just adding something to the velocity components so the two should be equivalent. A source term might allow a bit more flexibility to play with the equations numerically.

**Comment by christophernhill** [comment]
The script level for a simple barotropic gyre with multiple levels would then look like

```
# # Barotropic double gyre with multiple vertical points but constant density
#
# This script simulates a Barotropic double gyre forced by a sinusoidl wind stress
# with several vertical levels using `ClimateMachine.Ocean.HydrostaticBoussinesqSuperModel`.

using Printf
using Plots
using ClimateMachine

ClimateMachine.init()

using ClimateMachine.Ocean
using ClimateMachine.Ocean.CartesianDomains

using ClimateMachine.GenericCallbacks: EveryXSimulationTime
using ClimateMachine.Ocean: steps, Δt, current_time
using CLIMAParameters: AbstractEarthParameterSet, Planet

# We begin by specifying the domain and mesh,

Lx=4e6
Ly=6e6
H=-3000
domain = CartesianDomain(
    elements = (10, 15, 2),
    polynomialorder = 4,
    polynomialorder = 2,
    x = (0, Lx),
    y = (0, Ly),
    z = (H, 0),
    periodicity = (false, false, false),
    boundary = ((1, 1), (1, 1), (2, 3)),
)

# Note that the default solid-wall boundary conditions are free-slip and
# insulating on tracers. Next, we specify model parameters and the sheared
# initial conditions

struct ParamSet <: AbstractEarthParameterSet end

initial_conditions = InitialConditions(
    u = (x, y, z) -> -0,
    v = (x, y, z) -> -0,
    θ = (x, y, z) -> 20.,
    # θ(x,y,z) = ( 20. * ( 1 - z/H ) ),
)

ρₒ=1000
τₒ=1e-1
wind_stress = (
    τ¹¹ = ( x, y, t, p) -> (τₒ / ρₒ) * cos(2 * y * π / Lʸ),
)
# wind_stress = ()


model = Ocean.HydrostaticBoussinesqSuperModel(
    domain = domain,
    time_step = 30,
    initial_conditions = initial_conditions,
    parameters = ParamSet(),
    turbulence_closure = (νʰ = 5e3, νᶻ = 5e-3, κʰ = 1e3, κᶻ = 1e-4),
    rusanov_wave_speeds = (cʰ = 1, cᶻ = 1.e-3),
    buoyancy = ( αᵀ = 0., ρₒ=ρₒ ),
    coriolis = (f₀=1.e-4, β=1.e-11),
    boundary_conditions = (
      OceanBC(Impenetrable(NoSlip()), Insulating()),
      OceanBC(Impenetrable(FreeSlip()), Insulating()),
      OceanBC(Penetrable(KinematicStress(stress=wind_stress)), Insulating()),
    ),
)

```

**Comment by blallen** [comment]
> As a relatively quick solution we can solve this by
> 
> 1. Adding `x, y, z` as auxiliary variables (hopefully in the future we could figure out how to infer `x, y, z` without this) and
> 2. Redefining `KinematicStress` to
> 
> ```julia
> struct KinematicStress{S} <: VelocityDragBC
>     stress :: S
> end
> ```
> 
> We then need
> 
> ```julia
> @inline function ocean_velocity_boundary_state!(
>     nf::NumericalFluxSecondOrder,
>     bc_velocity::Impenetrable{KinematicStress},
>     ocean,
>     Q⁺,
>     D⁺,
>     A⁺,
>     n⁻,
>     Q⁻,
>     D⁻,
>     A⁻,
>     t,
> )
>     Q⁺.u = Q⁻.u
>     D⁺.ν∇u = n⁻ * bc_velocity.drag(A⁻.x, A⁻.y, A⁻.z, t)
> 
>     return nothing
> end
> ```
> 
> One then constructs a `KinematicStress` object with a function that returns the same object that `kinematic_stress(ocean.problem, A⁻.y, ocean.ρₒ)` would have returned.
> 
> As a side note, I think it is inconvenient that users need to specify their stress as a vector here. I think it would be more convenient to specify the components of the stress vector independently.

This is what @simonbyrne has done for the atmosphere and what we should do as a first step. I have some stub branch somewhere where I started on this but didn't get past GPU debugging. This would also let us keep all the parameters for the boundary conditions within the boundary condition struct instead of needing to be in the problem, model, or somewhere else.

**Comment by blallen** [comment]
> @christophernhill suggests that an alternative method for diagnosing the boundary you're currently on would be to rewrite this function:
> 
> https://github.com/CliMA/ClimateMachine.jl/blob/84caa94eb1bfb7ffd48cef73a8a400b70fc7d87f/src/Ocean/OceanBC.jl#L97-L130
> 
> I'm not sure how to do that; I would need someone to explain to me what this function is doing.

This `@generated` approach is being moved to the kernels as the standard method for unrolling the physical boundaries. see #1727 and the discussion there for details.

**Comment by blallen** [comment]
> we can also write a generic components_to_tensor function I think to fill out
> 
> ```
> D⁺.ν∇u
> ```
> 
> etc.... from generic stress components τ¹¹, τ²² relative to the local boundary. Might be best to start with simple case - rectangular, orthogonal grid and only diagonals for only tangential to wall vectors.

ν∇u is a 2x3 tensor with the following components: `[[t_11, t_12, t_13], [t_21, t_22, t_23]]`. The current term takes the transpose of the kinematic stress before multiplying with the normal vector, closer to an outer product than an inner product:
https://github.com/CliMA/ClimateMachine.jl/blob/25f4421449c0e3f10366b7ffb540992127301101/src/Ocean/HydrostaticBoussinesq/bc_velocity.jl#L233

**Comment by blallen** [comment]
> So I think the form for stress could be along the lines of
> 
> ```
>    D⁺.ν∇u = n⁻ * (@SVector[ bc_velocity.stress.τ¹¹( x,y,z,t),bc_velocity.stress.τ²²(x,y,z,t)])'
> ```
> 
> with τ¹¹ and τ²² defined in the scripting layer and -0 by default
> 
> rather than
> 
> ```
>     D⁺.ν∇u = n⁻ * kinematic_stress(ocean.problem, A⁻.y, ocean.ρₒ)'
> ```
> 

I guess the question here is how coupled the `x` and `y` components of the kinematic stress are. 

> A couple of things I am unclear on - but don't stop coding
> 
> * its a little fuzzy to me why D⁺.ν∇u seems to be a 2 element vector and not a full gradient tensor thing.

I think I answered this in the comment above, but we take the transpose of the kinematic stress and that results in the tensor.

> * we currently seem to set the bc in `nf::NumericalFluxSecondOrder,`, which is because we are setting through a k.grad term at boundary. Thats OK, but we could equally have the k.grad term as zero I think and have some sort of source term? In the end we are just adding something to the velocity components so the two should be equivalent. A source term might allow a bit more flexibility to play with the equations numerically.

I think this might introduce conservation problems? Handling it at the numerical flux level gets you a lot of nice things automatically, but we can ask @jkozdon and @lcw about that.

**Comment by blallen** [comment]
> > I also definitely agree with having a layer where we define Dirichlet, Neumann and Robin BCs.
> 
> Robin is another kind of BC that I didn't mention. That'd be good too.
> 
> Specifying fluxes directly should also be an option, I think, though the precise application to DG methods is hazy to me. But as far as I can tell, specifying fluxes is the simplest boundary condition in the weak formulation of the equations (?) For large-scale ocean / atmospheric modeling (LES to GCM scales) it seems that this is in fact the only boundary condition we need. Idealized work does benefit from the ability to prescribe values (Dirichlet) and gradients (Neumann) on boundaries... so we should support those too.

Specifying fluxes is the exact same thing as specifying `D⁺.ν∇u` in this case. I actually reverse-engineered this implementation from that of the `numerical_boundary_flux_second_order!` so that the `kinematic_stress()` is the flux applied on the boundary. Take a look at https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Numerics/DGMethods/NumericalFluxes.jl#L586-L632 for the details. You can also check out the numerics design docs for the motivation behind this specification.

The only case that isn't handled by this method is the Dirichlet case, which is rather straightforward as you just need to assign the boundary value as the plus side value and then not touch the other states.

**Comment by glwagner** [comment]
> I guess the question here is how coupled the `x` and `y` components of the kinematic stress are.

In idealized work (where this might matter most) I think it's common to align your stress with the coordinate system. For example, the double gyre prescribes stress in x only.

For realistic cases certainly the stress has both components, but in that case we have to somehow ingest a dataset and programatically prescribe boundary conditions rather than specifying them as functions in a user script.

**Comment by christophernhill** [comment]
> > I guess the question here is how coupled the `x` and `y` components of the kinematic stress are.
> 
> In idealized work (where this might matter most) I think it's common to align your stress with the coordinate system. For example, the double gyre prescribes stress in x only.
> 
> Does this depend on whether our equations are coordinate free, though? I think its ultimately a minor sacrifice to bundle the components if we get coordinate free notation out of it, somehow...
> 
> For realistic cases certainly the stress has both components, but in that case we have to somehow ingest a dataset and programatically prescribe boundary conditions rather than specifying them as functions in a user script.

I think even in realistic case, (or coupled model case) the scripting functions can be used as building blocks. They might call out to some pre-existing code that is provided to stage things (including in some GPU buffer as needed) from elsewhere etc.... - but same general abstraction could still work I think. It would need some code to provide an example of that at some point, but then people would see what to do.

**Comment by glwagner** [comment]
Quite agree! I think what I'm trying to say is that its likely we'll provide tools for this in source, which means imposing realistic bcs / bcs from data may not impose requirements on the user interface. We may want to consider how to make idealized setup easy to drive interface design?

**Comment by jkozdon** [comment]
> I think this might introduce conservation problems? Handling it at the numerical flux level gets you a lot of nice things automatically, but we can ask @jkozdon and @lcw about that.

Yes the traditional way to enforce boundary conditions DG is through the numerical fluxes. There are probably other ways that could be formulated, but as brandon pointed out you get conservation with numerical fluxes and also you are penalizing the solution to be in a certain space of functions.
